spec/lib/rex/proto/ldap/server_spec.rb

```diff
@@ -15,7 +15,32 @@
 
   let(:response) {}
 
+  let(:tcp_server_socket) do
+    double :tcp_server_socket,
```

Comment:
It's not possible to use `instance_double Rex::Socket::TcpServer` here for an InstanceVerifyingDouble - as behind the scenes in rex-text a Socket instance is actually returned that's been monkey patched with modules like Rex::Socket::TcpServer etc, which makes it awkward to use `instance_double` here

```
[5] pry(#<Rex::Proto::LDAP::Server>)> whereami

From: /Users/user/Documents/code/metasploit-framework/lib/rex/proto/ldap/server.rb:102 Rex::Proto::LDAP::Server#start:

     97:           end
     98: 
     99:           if serve_tcp
    100:             @tcp_sock = Rex::Socket::TcpServer.create(sock_options)
    101:             require 'pry-byebug'; binding.pry
 => 102:             tcp_sock.on_client_connect_proc = proc do |cli|
    103:               on_client_connect(cli)
    104:             end
    105:             tcp_sock.on_client_data_proc = proc do |cli|
    106:               on_client_data(cli)
    107:             end

[6] pry(#<Rex::Proto::LDAP::Server>)> @tcp_sock.class
=> Socket
[7] pry(#<Rex::Proto::LDAP::Server>)> @tcp_sock.singleton_class.included_modules
=> [Rex::Socket::TcpServer,
 Rex::IO::StreamServer,
 Rex::Socket,
 Net::BER::BERParser,
  ...]
```

Reply:
If possible can you explain why that was happening?

Reply:
For unit tests we generally want to avoid creating real servers etc. So I've gone with mocking the `Rex::Socket::TcpServer.create(...)` call.

The rex-socket implementation uses Ruby's API for creating a socket here:

https://github.com/rapid7/rex-socket/blob/fba62cdf31e2804b91667b34759a31da823bbeb7/lib/rex/socket/tcp_server.rb#L36-L40

And here:

https://github.com/rapid7/rex-socket/blob/fba62cdf31e2804b91667b34759a31da823bbeb7/lib/rex/socket/comm/local.rb#L170-L175

Then adds additional methods and state onto the socket instance with `.extend ...` here:

https://github.com/rapid7/rex-socket/blob/fba62cdf31e2804b91667b34759a31da823bbeb7/lib/rex/socket/tcp_server.rb#L54

Let me know if I haven't answered your question 😄